### ISSUE TITLE
Kitty URL Handler for ssh/telnet or other application on macOS

### ISSUE THREAD

**Issue by ciscohack** [kovidgoyal/kitty#4921]
Hello Expert,

I want to use my kitty as URL handler for telnet/ssh based URL to access my lab devices. i checked the documentation where URL handler config is mentioned for many protocol. below is the config in that telnet is missing second i put this config but it seems not working.

"**url_prefixes file ftp ftps gemini git gopher http https irc ircs kitty mailto news sftp ssh**" 

I have also checked open-controls.conf option also tried as mentioned in documentation 

this is my open-actions.conf setting 
# Open ssh URLs with telnet command
protocol telnet
action launch --type=os-window telnet $URL

above setting also not working and whenever i click on URL like **telnet://192.168.1.1 : 35402**  to open the device in kitty app. Kitty terminal complain saying below error

Open URL Error

Unknown URL type, cannot open:

how can i fix please help me if this is already doable .. if not doable can we include this in future releases.

**Comment by page-down** [comment]
You first need to understand that there is a difference between "clicking a link in kitty" and "clicking a link in another macOS app and expecting it to be opened by kitty".

> ... it seems not working ...
> url_prefixes ...

This option allows some links to be recognized and clickable in kitty.

> ... open-actions.conf ...

This defines the action when the link is clicked in kitty.

What you need is `launch-actions.conf`, read the documentation again.
Don't forget to register certain protocols to be opened by kitty.

https://sw.kovidgoyal.net/kitty/open_actions/#scripting-the-opening-of-files-with-kitty-on-macos

**Comment by ciscohack** [comment]
@page-down thanks for the respond. I tried that launch-actions.conf file option and put below config 

# Open telnet URLs with ssh command
protocol telnet
action launch --type=os-window telnet $URL

with open-actions.conf i was able to open kitty app but url handler was failing but with launch action nothing working.. Can you please give me a sample configuration snippet which i can try to achieve telnet/ssh url handler to clock on any telnet based URL which further get open in kitty app and access the device over telnet protocol

**Comment by page-down** [comment]
I suggest you solve this problem by following these steps.

- Do I set the default telnet URL handler to kitty? How do I know the current telnet URL's default app?
- What kind of parameters does the `telnet` command line accept? Does `telnet telnet://example.org` work? How do I check the telnet man page?
- What kind of string does the `$URL` configured in kitty `launch-actions.conf` convert to? How can I confirm this myself? If there is no variable to get hostname directly, how can I convert it to hostname?
- After modifying `launch-actions.conf` I found that it doesn't work. Did I try to exit kitty.app completely and then re-open it?

When you have studied all the above questions and got clear answers, I think this problem will be solved.

**Comment by ciscohack** [comment]
@page-down 
=> Do I set the default telnet URL handler to kitty? How do I know the current telnet URL's default app?
telnet url IP will be same sometime but port number will be random. URL will be always look like this format
( **telnet://192.168.1.1 : 35402**)

=> What kind of parameters does the telnet command line accept? Does telnet telnet://example.org work? How do I check the telnet man page?

There is no as such parameter i going to pass. it just be IP : Port only as i mentioned in 1st point. man page i checked and tried action launch like this (action launch --type=os-window telnet $URL $$HOST$$ $$PORT$$) but no luck failing.

=> What kind of string does the $URL configured in kitty launch-actions.conf convert to? How can I confirm this myself? If there is no variable to get hostname directly, how can I convert it to hostname?

why do we need to convert hostname when i am accessing the device through IP and device also IP based

=> After modifying launch-actions.conf I found that it doesn't work. Did I try to exit kitty.app completely and then re-open it?

Yes i did app restart after change but not working 


Kitty as terminal is best no doubt .. performance is killer document is also very good but some how some basic stuff configuration sometime feel need more expertise on terminal config :P 

I can see open-actions.conf/launch-action.conf  i tried what mentioned in doc but success rate is 1 or 0%.

It would be good one small demo video where we can show open-actions.config .. config and immediate effect what config will outcome to get better understanding ...

Like i tried this config in open-actions.conf

# Open directories
protocol file
mime inode/directory
action launch --type=os-window --cwd $FILE_PATH

I suppose i should get hyperlink for my ls and when i click i should enter in directory but sadly this also not working :)

URL handler is very basic feature in all terminal and it working fine for me on other terminal . it just i switched to Kitty and i loved kitty not want to go back but noticed this feature is not working on kitty  :(

**Comment by page-down** [comment]
>> Do I set the default telnet URL handler to kitty? 
>
> telnet url IP will be same sometime but port number ...

No, what I mean is that you need to get macOS to use kitty when opening a telnet URL.
Open the following page and search for `Set kitty as the handler for ssh:// URLs`.
https://sw.kovidgoyal.net/kitty/open_actions/

Copy the code and set the telnet URL to open with kitty.
```shell
kitty +runpy '# ... copy codes to here ...' telnet
```

>> ... Does `telnet telnet://example.org` work? ...
>
> ...  i checked and tried action launch like this (action launch --type=os-window telnet $URL $$HOST$$ $$PORT$$) but no luck failing.

>> What kind of string does the $URL configured in kitty launch-actions.conf convert to? ...
>
> why do we need to convert hostname when i am accessing the device through IP ...

What I mean here is that telnet needs the HOSTNAME (IP) part of the URL.
You should have read the kitty documentation (link above) and it never mentions `$$HOST$$`, `$$PORT$$`, so you need to convert it yourself.
You can't just ignore the documentation, fill in `$$HOST$` or any random name and expect it to work.
So, here you need a wrapper script that converts `$URL` and then calls `telnet <IP_HERE> <PORT_HERE>`.
In fact, when I implemented this feature, I thought about whether we needed to provide HOSTNAME (IP), PORT, etc.
However, this part of code was not committed and merged. It is up to kovid to decide whether these need to be added or not.

> ... but some how some basic stuff configuration sometime feel need more expertise  ...
> It would be good one small demo video ...
> ... not want to go back but noticed this feature is not working on kitty ...

Unfortunately kitty does require a certain understanding of UNIX and scripting, and there are not many hands-on video tutorials available.

> Like i tried this config in open-actions.conf ...
> ... when i click i should enter in directory but sadly this also not working ...

Still sorry to say that I tried the exact same configuration as you did, however it worked smoothly and normally.
I don't know what part of your environment went wrong, it would be for you to sort it out, for example starting from scratch.

**Comment by ciscohack** [comment]
@page-down sorry i forgot to mention yes i have setup in my macOS kitty.app as default app for ssh/telnet and terminal operation with the help of defaultapp tool

**Comment by ciscohack** [comment]
@page-down This veriable $$HOST$$, $$PORT$$ i just tried to see if that help solving my purpose but initially i setup config as per document infact i have shared my config 

```
 Open script files based on their shebangs
protocol file
ext sh,command,tool
action launch --hold --type=os-window kitty +shebang $FILE_PATH {SHELL}

# Open shell specific script files
protocol file
ext fish,bash,zsh
action launch --hold --type=os-window kitty +shebang $FILE_PATH __ext__

# Open directories
protocol file
mime inode/directory
action launch --type=os-window --cwd $FILE_PATH

# Open text files without fragments in the editor
protocol file
mime text/*
action launch --type=os-window $EDITOR $FILE_PATH

# Open image files with icat
protocol file
mime image/*
action launch --type=os-window zsh -c "export PATH=/opt/homebrew/bin:/opt/homebrew/sbin:$PATH; kitty +kitten icat --hold '$FILE_PATH'"

# Open ssh URLs with ssh command
protocol ssh
action launch --type=os-window ssh $URL

# Open telnet URLs with ssh command
 protocol telnet
action launch --type=overlay telnet  ssh $URL
```
This is my launch-action.conf file

**Comment by ciscohack** [comment]
@page-down yeah after using kitty i have noticed this need some unix/scripting skill.. unfortunately i am zero.. only know some basic linux stuff for day to day operations.
